### ISSUE TITLE
Bug: Mixed image extensions in WebDataset lead to silent data loss

### ISSUE THREAD

**Issue by hongwqin** [Amshaker/Mobile-O#8]
Hi, thanks for open-sourcing this great project.

I ran into what looks like a data-loading issue during SFT training and wanted to report it here in case it is helpful.

#### Description

When loading multiple WebDataset .tar files with mixed image extensions (e.g., .jpg in one tar and .png in another), the Hugging Face datasets loader may infer inconsistent schemas.

As a result, during the image column normalization in:

https://github.com/Amshaker/Mobile-O/blob/520d1504340100a1626678db4b9890924ae6e90d/mobileo/train/train.py#L423-L432

only one image field (e.g., jpg) is preserved, while others (e.g., png) become None. After unused columns are removed, the actual image data is lost.

This leads to runtime warnings like:

[WARN] Skipping corrupted sample: 'NoneType' object has no attribute 'convert'

and silently drops a large portion of training samples.

#### Expected behavior

All image formats across tar files should be correctly mapped to a unified image field without data loss.

#### Actual behavior
Mixed extensions → inconsistent schema
Some samples get image=None
Training skips these samples due to .convert("RGB") failure

#### Impact
Silent data loss
Noisy training logs
Potential degradation in training quality

#### Suggested fix

Process each .tar file independently before concatenation, and align schema per file (e.g., normalize image column names before merging). Here is the fix PR https://github.com/Amshaker/Mobile-O/pull/7

**Comment by Amshaker** [comment]
Hi @hongwqin ,

Thank you for sharing your feedback and for the PR.

However, I wasn't able to reproduce the issue on my end. I double-checked and don't see this warning in my setup. The total number of examples printed in the training log is:

`Num examples = 104,576`

which is the expected count (59K from BLIP3o + 45K from ShareGPT). Could you share more details about your environment or configuration? I believe it might be a folder structure issue rather than an extension issue.

**Comment by hongwqin** [comment]
Hi @Amshaker 
Thanks for checking. I also ran a validation script on my side, and the total sample count is indeed correct. The problem is not that samples are missing, but that the image field becomes None for a large subset of samples after the column rename step. The script I used is attached here .  

[check_dataset.py](https://github.com/user-attachments/files/26664129/check_dataset.py)

The running command and result is here:

```bash
python check_dataset.py --data_path "nvme/mobile-o-data/Mobile-O-SFT" 
```

<img width="2307" height="502" alt="Image" src="https://github.com/user-attachments/assets/025fd554-9dd2-4822-b540-83e46d8046c1" />

From my run, the dataset loads 104,644 samples in total. However, after the current loading and renaming logic, 58,859 samples end up with image is None (about 56.25%), even though the sample count itself is unchanged.

So the issue is specifically in how image columns are unified. Some tar files store images under jpg, while others use png. When all tar files are loaded together, the dataset may contain both kinds of image fields. The current code then assumes there is only one global image column and does:

```python
train_dataset = train_dataset.rename_column("png", "image")
```

or the reverse for `jpg`.

This causes samples whose actual image data is stored in the other column to get `image = None`. Their image bytes are still present in the original column at that point, but that column is later removed, so the image is effectively dropped.

In other words, the total number of examples is correct, but many samples become corrupted because their image data is discarded during the rename + cleanup step.

I believe the fix should be to handle each tar file separately before concatenation, or explicitly merge `jpg/png` into a single `image` field without dropping whichever one is not renamed.

**Comment by Amshaker** [comment]
Hi @hongwqin ,

Thank you for your detailed analysis!

I figured out that the extension of the data is unified in my setup, and your solution actually fixes the issue for the mixed extensions.

I merged your pull request, and thank you again for your detailed analysis.
